src/Microsoft.ML.CpuMath/FactorizationMachine/FmaIntrinsics.cs

```diff
@@ -0,0 +1,188 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace Microsoft.ML.Internal.CpuMath.FactorizationMachine
+{
+    internal static class FmaIntrinsics
```

Comment:
Since this class is nearly identical to the `AvxIntrinsics` class, would it make more sense to collapse these two classes together, and then just interleave `if (Fma.IsSupported)` inline in the methods when you want to switch use an `Fma` intrinsic?

That way there isn't so much duplicated code.

We already follow that pattern here: [https://github.com/dotnet/machinelearning/blob/fabdabf867dc68481d519740a96b981bbeb63180/src/Microsoft.ML.CpuMath/AvxIntrinsics.cs#L159-L171](https://github.com/dotnet/machinelearning/blob/fabdabf867dc68481d519740a96b981bbeb63180/src/Microsoft.ML.CpuMath/AvxIntrinsics.cs#L159-L171)

Reply:
I agree. We can interleave in the methods to avoid code duplication. I separated the implementation into two classes to make it more readable. I will update the code to reflect these changes.